File: problem_6/problem_6.js
// The sum of the squares of the first ten natural numbers is,
// 1^2 + 2^2 + ... + 10^2 = 385
// The square of the sum of the first ten natural numbers is,
// (1 + 2 + ... + 10)^2 = 55^2 = 3025
// Hence the difference between the sum of the squares of the first ten natural numbers and the square of the sum is 3025 − 385 = 2640.
// Find the difference between the sum of the squares of the first one hundred natural numbers and the square of the sum.

const now = require("performance-now");

// iterative method
const problem6a = n => {
    let sumOfSquares = 0;
    let squareOfSum = 0;
    
    for (var i = 1; i <= n; i++){
        sumOfSquares += i**2;
        squareOfSum += i;
    }
    return squareOfSum**2 - sumOfSquares;
}
// mathematical method
const problem6b = n => {
    let sumOfSquares = n * (n + 1) * (2*n + 1) * (1 / 6);
    let squareOfSum = (n * (n + 1) / 2) ** 2;
    return squareOfSum - sumOfSquares;
}

let start1 = now();
let solution1 = problem6a(100)
console.log(`${solution1} found in ${(now() - start1) / 1000} seconds.`)
// 25164150 found in 0.00003263299999999702 seconds.

let start2 = now();
let solution2 = problem6b(100)
console.log(`${solution2} found in ${(now() - start2) / 1000} seconds.`)
// 25164150 found in 0.00006033100000000502 seconds.

File: problem_6/problem_6.py
# The sum of the squares of the first ten natural numbers is,
# 1^2 + 2^2 + ... + 10^2 = 385
# The square of the sum of the first ten natural numbers is,
# (1 + 2 + ... + 10)^2 = 55^2 = 3025
# Hence the difference between the sum of the squares of the first ten natural numbers and the square of the sum is 3025 − 385 = 2640.
# Find the difference between the sum of the squares of the first one hundred natural numbers and the square of the sum.

import time

# iterative method
def problem_6a(n):
    sum_of_squares = 0
    square_of_sum = 0
    
    for i in range(n + 1):
        sum_of_squares += i**2
        square_of_sum += i

    return square_of_sum**2 - sum_of_squares

# mathematical method
def problem_6b(n):
    sum_of_squares = n * (n + 1) * (2*n + 1) / 6
    square_of_sums = (n * (n + 1) / 2) ** 2

    return square_of_sums - sum_of_squares    

start = time.time()
solution = problem_6a(100)
print(f"{solution} found in {time.time() - start} seconds.")
# 25164150 found in 3.981590270996094e-05 seconds.

start = time.time()
solution = problem_6b(100)
print(f"{solution} found in {time.time() - start} seconds.")
# 25164150.0 found in 1.3113021850585938e-05 seconds.